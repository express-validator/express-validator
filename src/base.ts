import { ReadonlyContext } from './context';

export interface Request {
  [k: string]: any;
  body?: any;
  cookies?: Record<string, any>;
  headers?: Record<string, any>;
  params?: Record<string, any>;
  query?: Record<string, any>;
}

export type Middleware = (req: Request, res: any, next: (err?: any) => void) => void;

export type Location = 'body' | 'cookies' | 'headers' | 'params' | 'query';

/**
 * Metadata about a validated field.
 */
export type Meta = {
  /**
   * The express request from which the field was validated
   */
  req: Request;

  /**
   * Which of the request objects the field was picked from
   */
  location: Location;

  /**
   * The full path of the field within the request object.
   *
   * @example
   * const meta = { req, location: 'body', path: 'foo.bar' }; // req.body.foo.bar
   */
  path: string;
};

/**
 * A function which may
 * - return falsy values, a promise that rejects or throw to indicate that a field is invalid;
 * - return truthy values or a promise that resolves to indicate that a field is valid.
 *
 * @param input the field value
 * @param meta metadata about the field being validated
 */
export type CustomValidator = (input: any, meta: Meta) => any;
export type StandardValidator = (input: string, ...options: any[]) => boolean;

export type CustomSanitizer = (input: any, meta: Meta) => any;
export type StandardSanitizer = (input: string, ...options: any[]) => any;

export interface FieldInstance {
  path: string;
  originalPath: string;
  location: Location;
  value: any;
  originalValue: any;
}

export type FieldValidationError = {
  /**
   * Indicates that the error occurred because a field had an invalid value
   */
  type: 'field';

  /**
   * The location within the request where this field is
   */
  location: Location;

  /**
   * The path to the field which has a validation error
   */
  path: string;

  /**
   * The value of the field
   */
  value: any;

  /**
   * The error message
   */
  msg: any;
};

export type AlternativeValidationError = {
  /**
   * Indicates that the error occurred because all alternatives (e.g. in `oneOf()`) were invalid
   */
  type: 'alternative';

  /**
   * The error message
   */
  msg: any;

  /**
   * The list of underlying validation errors returned by validation chains in `oneOf()`
   */
  nestedErrors: FieldValidationError[];
};

/**
 * A validation error as reported by a middleware.
 * The properties available in the error object vary according to the type.
 *
 * @example
 *  if (error.type === 'alternative') {
 *    console.log(`There are ${error.nestedErrors.length} errors under this alternative list`);
 *  } else if (error.type === 'field') {
 *    console.log(`There's an error with field ${error.path) in the request ${error.location}`);
 *  }
 *
 */
export type ValidationError = AlternativeValidationError | FieldValidationError;

/**
 * A function which creates an error message based on a field's value.
 *
 * @param input the field value
 * @param meta metadata about the field that was validated
 */
export type FieldMessageFactory = (value: any, meta: Meta) => any;

/**
 * A function which creates an error message based on an alternative's nested errors.
 *
 * @see `oneOf()`
 * @param nestedErrors The errors from the invalid alternative(s).
 * @param opts
 */
export type AlternativeMessageFactory = (
  nestedErrors: FieldValidationError[],
  opts: { req: Request },
) => any;

// Not using Symbol because of #813
export const contextsKey = 'express-validator#contexts';

export interface InternalRequest extends Request {
  [contextsKey]?: ReadonlyContext[];
}

export class ValidationHalt extends Error {}
