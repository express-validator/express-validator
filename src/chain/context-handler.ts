import { CustomValidator } from '../base';
import { Optional } from '../context';
import { ContextRunner } from './context-runner';

export interface BailOptions {
  /**
   * Defines the level at which to stop running further validations:
   * - When set to `chain`, further validations won't be run for this validation chain if there
   *   are any errors.
   * - When set to `request`, no further validations on the same request will be run either if
   *   there are any errors.
   *
   * @default 'chain'
   */
  level?: 'chain' | 'request';
}

export interface ContextHandler<Chain> {
  /**
   * Stops running validations if any of the previous ones have failed.
   *
   * Useful to prevent a custom validator that touches a database or external API from running when
   * you know it will fail.
   *
   * May be used multiple times in the same validation chain if desired.
   *
   * @example
   *  check('username')
   *    .isEmail()
   *    // If not an email, stop here
   *    .bail()
   *    .custom(checkDenylistDomain)
   *    // If domain is not allowed, don't go check if it already exists
   *    .bail()
   *    .custom(checkEmailExists)
   *
   * @returns the current validation chain
   */
  bail(opts?: BailOptions): Chain;

  /**
   * Adds a condition on whether the validation should continue on a field or not.
   * @param condition may be either
   *  - a custom validator-like function, which must truthy or a promise that resolves to continue
   *    validation. If the return value is falsy, a promise that rejects, or if it throws, validation
   *    will stop.
   * - a validation chain which if it would produce errors, the validation chain stops.
   * @example
   *  check('newPassword')
   *    // Only validate if the old password has been provided
   *    .if((value, { req }) => req.body.oldPassword)
   *    // Or, use it with a a validation chain
   *    .if(body('oldPassword').notEmpty())
   * @returns the current validation chain
   */
  if(condition: CustomValidator | ContextRunner): Chain;

  /**
   * Marks the field(s) of the validation chain as optional.
   * By default, only fields with an `undefined` value are considered optional and will be ignored
   * when validating.
   *
   * @param options an object of options to customize the behavior of optional.
   * @returns the current validation chain
   */
  optional(options?: Partial<Optional> | true): Chain;
}
