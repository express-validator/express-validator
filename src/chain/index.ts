export * from './sanitizers';
export * from './sanitizers-impl';
export * from './context-handler';
export * from './context-handler-impl';
export * from './context-runner';
export * from './context-runner-impl';
export * from './validators';
export * from './validators-impl';
export * from './validation-chain';
