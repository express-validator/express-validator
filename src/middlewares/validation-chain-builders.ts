import { ErrorMessage, FieldMessageFactory, Location } from '../base';
import { check as baseCheck } from './check';

/**
 * Creates a variant of `check()` that checks the given request locations.
 *
 * @example
 *  const checkBodyAndQuery = buildCheckFunction(['body', 'query']);
 */
export function buildCheckFunction(locations: Location[]) {
  return (fields?: string | string[], message?: FieldMessageFactory | ErrorMessage) =>
    baseCheck(fields, locations, message);
}

/**
 * Creates a middleware/validation chain for one or more fields that may be located in
 * any of the following:
 *
 * - `req.body`
 * - `req.cookies`
 * - `req.headers`
 * - `req.params`
 * - `req.query`
 *
 * @param fields  a string or array of field names to validate/sanitize
 * @param message an error message to use when failed validations don't specify a custom message.
 *                Defaults to `Invalid Value`.
 */
export const check = buildCheckFunction(['body', 'cookies', 'headers', 'params', 'query']);

/**
 * Same as {@link check()}, but only validates `req.body`.
 */
export const body = buildCheckFunction(['body']);

/**
 * Same as {@link check()}, but only validates `req.cookies`.
 */
export const cookie = buildCheckFunction(['cookies']);

/**
 * Same as {@link check()}, but only validates `req.headers`.
 */
export const header = buildCheckFunction(['headers']);

/**
 * Same as {@link check()}, but only validates `req.params`.
 */
export const param = buildCheckFunction(['params']);

/**
 * Same as {@link check()}, but only validates `req.query`.
 */
export const query = buildCheckFunction(['query']);
