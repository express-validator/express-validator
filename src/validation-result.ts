import * as _ from 'lodash';
import { InternalRequest, Request, ValidationError, contextsKey } from './base';
import { bindAll } from './utils';

/**
 * Given a validation error, returns a new value that represents it.
 */
export type ErrorFormatter<T = any> = (error: ValidationError) => T;
type ToArrayOptions = {
  /**
   * Whether only the first error of each field should be returned.
   * @default false
   */
  onlyFirstError?: boolean;
};

// eslint-disable-next-line no-use-before-define
export type ResultFactory<T> = (req: Request) => Result<T>;

interface ResultFactoryBuilderOptions<T = any> {
  /**
   * The default error formatter of every {@link Result} instance returned by
   * the custom `validationResult()` function.
   */
  formatter: ErrorFormatter<T>;
}

/**
 * Extracts the validation errors of an express request
 */
export const validationResult = Object.assign(withDefaults<ValidationError>(), { withDefaults });

/**
 * The current state of the validation errors in a request.
 */
export class Result<T = any> {
  constructor(
    private formatter: ErrorFormatter<T>,
    private readonly errors: readonly ValidationError[],
  ) {}

  /**
   * Gets the validation errors as an array.
   *
   * @param options.onlyFirstError whether only the first error of each
   */
  array(options?: ToArrayOptions): T[] {
    return options && options.onlyFirstError
      ? Object.values(this.mapped())
      : this.errors.map(this.formatter);
  }

  /**
   * Gets the validation errors as an object.
   * If a field has more than one error, only the first one is set in the resulting object.
   *
   * @returns an object from field name to error
   */
  mapped(): Record<string, T> {
    return this.errors.reduce((mapping, error) => {
      const key = error.type === 'field' ? error.path : `_${error.type}`;
      if (!mapping[key]) {
        mapping[key] = this.formatter(error);
      }

      return mapping;
    }, {} as Record<string, T>);
  }

  /**
   * Specifies a function to format errors with.
   * @param formatter the function to use for formatting errors
   * @returns A new {@link Result} instance with the given formatter
   */
  formatWith<T2>(formatter: ErrorFormatter<T2>): Result<T2> {
    return new Result(formatter, this.errors);
  }

  /**
   * @returns `true` if there are no errors, `false` otherwise
   */
  isEmpty() {
    return this.errors.length === 0;
  }

  /**
   * Throws an error if there are validation errors.
   */
  throw() {
    if (!this.isEmpty()) {
      throw Object.assign(new Error(), bindAll(this));
    }
  }
}

/**
 * Creates a `validationResult`-like function with default options passed to every {@link Result} it
 * returns.
 */
function withDefaults<T = any>(
  options: Partial<ResultFactoryBuilderOptions<T>> = {},
): ResultFactory<T> {
  const defaults: ResultFactoryBuilderOptions<ValidationError> = {
    formatter: error => error,
  };
  const actualOptions = _.defaults(options, defaults);

  return (req: InternalRequest) => {
    const contexts = req[contextsKey] || [];
    const errors = _.flatMap(contexts, 'errors');
    return new Result(actualOptions.formatter, errors);
  };
}
