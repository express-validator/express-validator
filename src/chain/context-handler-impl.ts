import { ContextBuilder } from '../context-builder';
import { Optional } from '../context';
import { ChainCondition, CustomCondition } from '../context-items';
import { CustomValidator } from '../base';
import { Bail } from '../context-items/bail';
import { BailOptions, ContextHandler, OptionalOptions } from './context-handler';
import { ContextRunner } from './context-runner';

export class ContextHandlerImpl<Chain> implements ContextHandler<Chain> {
  constructor(private readonly builder: ContextBuilder, private readonly chain: Chain) {}

  bail(opts?: BailOptions) {
    if (opts?.level === 'request') {
      this.builder.setRequestBail();
    }
    this.builder.addItem(new Bail());
    return this.chain;
  }

  if(condition: CustomValidator | ContextRunner) {
    if ('run' in condition) {
      this.builder.addItem(new ChainCondition(condition));
    } else if (typeof condition === 'function') {
      this.builder.addItem(new CustomCondition(condition));
    } else {
      throw new Error('express-validator: condition is not a validation chain nor a function');
    }
    return this.chain;
  }

  optional(options: OptionalOptions | boolean = true) {
    let value: Optional;
    if (typeof options === 'boolean') {
      value = options ? 'undefined' : false;
    } else {
      value = options.values
        ? options.values
        : options.checkFalsy
        ? 'falsy'
        : options.nullable
        ? 'null'
        : 'undefined';
    }

    this.builder.setOptional(value);
    return this.chain;
  }
}
