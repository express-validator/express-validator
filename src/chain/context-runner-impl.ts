import * as _ from 'lodash';
import { InternalRequest, Request, ValidationHalt, contextsKey } from '../base';
import { Context, ReadonlyContext } from '../context';
import { ContextBuilder } from '../context-builder';
import { SelectFields, selectFields as baseSelectFields } from '../select-fields';
import { Result } from '../validation-result';
import { ContextRunner, ResultWithContext } from './context-runner';

export class ResultWithContextImpl extends Result implements ResultWithContext {
  constructor(readonly context: ReadonlyContext) {
    super(error => error, context.errors);
  }
}

export class ContextRunnerImpl implements ContextRunner {
  constructor(
    private readonly builderOrContext: ContextBuilder | Context,
    private readonly selectFields: SelectFields = baseSelectFields,
  ) {}

  async run(req: Request, options: { dryRun?: boolean } = {}) {
    const context =
      this.builderOrContext instanceof Context
        ? this.builderOrContext
        : this.builderOrContext.build();
    const instances = this.selectFields(req, context.fields, context.locations);
    context.addFieldInstances(instances);

    const haltedInstances = new Set<string>();

    for (const contextItem of context.stack) {
      const promises = context.getData({ requiredOnly: true }).map(async instance => {
        const { location, path } = instance;
        const instanceKey = `${location}:${path}`;
        if (haltedInstances.has(instanceKey)) {
          return;
        }

        try {
          await contextItem.run(context, instance.value, {
            req,
            location,
            path,
          });

          // An instance is mutable, so if an item changed its value, there's no need to call getData again
          const newValue = instance.value;

          // Checks whether the value changed.
          // Avoids e.g. undefined values being set on the request if it didn't have the key initially.
          const reqValue = path !== '' ? _.get(req[location], path) : req[location];
          if (!options.dryRun && reqValue !== instance.value) {
            path !== '' ? _.set(req[location], path, newValue) : _.set(req, location, newValue);
          }
        } catch (e) {
          if (e instanceof ValidationHalt) {
            haltedInstances.add(instanceKey);
            return;
          }

          throw e;
        }
      });

      await Promise.all(promises);
    }

    if (!options.dryRun) {
      const internalReq = req as InternalRequest;
      internalReq[contextsKey] = (internalReq[contextsKey] || []).concat(context);
    }

    return new ResultWithContextImpl(context);
  }
}
