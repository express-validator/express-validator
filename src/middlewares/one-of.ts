import * as _ from 'lodash';
import { ContextRunnerImpl, ValidationChain } from '../chain';
import { InternalRequest, Middleware, Request } from '../base';
import { ContextBuilder } from '../context-builder';
import { ContextItem } from '../context-items';
import { Result } from '../validation-result';

// A dummy context item that gets added to surrogate contexts just to make them run
const dummyItem: ContextItem = { async run() {} };

export type OneOfCustomMessageBuilder = (options: { req: Request }) => any;

export type OneOfErrorType = 'grouped' | 'leastErroredOnly' | 'flat';

// Not used in this file, just for third party users.
export type OneOfOptions =
  | {
      message?: OneOfCustomMessageBuilder;
      errorType?: OneOfErrorType;
    }
  | {
      message?: any;
      errorType?: OneOfErrorType;
    };

/**
 * Creates a middleware that will ensure that at least one of the given validation chains
 * or validation chain groups are valid.
 *
 * If none are, a single error for a pseudo-field `_error` is added to the request,
 * with the errors of each chain made available under the `nestedErrors` property.
 *
 * @param chains an array of validation chains to check if are valid.
 *               If any of the items of `chains` is an array of validation chains, then all of them
 *               must be valid together for the request to be considered valid.
 * @param options.message a function for creating a custom error message in case none of the chains are valid
 */
export function oneOf(
  chains: (ValidationChain | ValidationChain[])[],
  options?: { message?: OneOfCustomMessageBuilder; errorType?: OneOfErrorType },
): Middleware & { run: (req: Request) => Promise<Result> };

/**
 * Creates a middleware that will ensure that at least one of the given validation chains
 * or validation chain groups are valid.
 *
 * If none are, a single error for a pseudo-field `_error` is added to the request,
 * with the errors of each chain made available under the `nestedErrors` property.
 *
 * @param chains an array of validation chains to check if are valid.
 *               If any of the items of `chains` is an array of validation chains, then all of them
 *               must be valid together for the request to be considered valid.
 * @param options.message an error message to use in case none of the chains are valid
 */
export function oneOf(
  chains: (ValidationChain | ValidationChain[])[],
  options?: { message?: any; errorType?: OneOfErrorType },
): Middleware & { run: (req: Request) => Promise<Result> };

export function oneOf(
  chains: (ValidationChain | ValidationChain[])[],
  options: { message?: any; errorType?: OneOfErrorType } = {},
): Middleware & { run: (req: Request) => Promise<Result> } {
  let result: Result;
  const middleware = async (req: InternalRequest, _res: any, next: (err?: any) => void) => {
    const surrogateContext = new ContextBuilder().addItem(dummyItem).build();

    // Run each group of chains in parallel, and within each group, run each chain in parallel too.
    const promises = chains.map(async chain => {
      const group = Array.isArray(chain) ? chain : [chain];
      const results = await Promise.all(group.map(chain => chain.run(req, { dryRun: true })));
      const contexts = results.map(result => result.context);
      const groupErrors = _.flatMap(contexts, 'errors');

      // #536: The data from a chain within oneOf() can only be made available to e.g. matchedData()
      // if its entire group is valid.
      if (!groupErrors.length) {
        contexts.forEach(context => {
          surrogateContext.addFieldInstances(context.getData());
        });
      }

      return groupErrors;
    });

    try {
      const allErrors = await Promise.all(promises);
      const success = allErrors.some(groupErrors => groupErrors.length === 0);

      if (!success) {
        let error;
        switch (options.errorType) {
          case 'flat':
            error = _.flatMap(allErrors);
            break;
          case 'leastErroredOnly':
            let leastErroredIndex = 0;
            for (let i = 1; i < allErrors.length; i++) {
              if (allErrors[i].length < allErrors[leastErroredIndex].length) {
                leastErroredIndex = i;
              }
            }
            error = allErrors[leastErroredIndex];
            break;
          default:
            // grouped
            error = allErrors;
        }

        // Only add an error to the context if no group of chains had success.
        surrogateContext.addError({
          type: 'nested',
          req,
          message: options.message || 'Invalid value(s)',
          nestedErrors: error,
        });
      }

      // Final context running pass to ensure contexts are added and values are modified properly
      result = await new ContextRunnerImpl(surrogateContext).run(req);
      next();
    } catch (e) {
      next(e);
    }
  };

  const run = async (req: Request) => {
    return new Promise<Result>((resolve, reject) => {
      middleware(req, {}, (e?: any) => {
        e ? reject(e) : resolve(result);
      });
    });
  };

  return Object.assign(middleware, { run });
}
