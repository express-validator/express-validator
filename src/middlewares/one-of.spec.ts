import { InternalRequest, contextsKey } from '../base';
import { ContextRunnerImpl } from '../chain/context-runner-impl';
import { Result } from '../validation-result';
import { check } from './validation-chain-builders';
import { OneOfErrorType, OneOfOptions, oneOf } from './one-of';

const getOneOfContext = (req: InternalRequest) => {
  const contexts = req[contextsKey] || [];
  return contexts[contexts.length - 1];
};

it('runs chains passed to it as a dry-run', done => {
  const req = {};
  const spy = jest.spyOn(ContextRunnerImpl.prototype, 'run');
  oneOf([check('bar'), check('baz')])(req, {}, () => {
    expect(spy).toHaveBeenCalledTimes(3);
    // Call 3 asserted by next test
    expect(spy).toHaveBeenNthCalledWith(1, req, { dryRun: true });
    expect(spy).toHaveBeenNthCalledWith(2, req, { dryRun: true });

    spy.mockRestore();
    done();
  });
});

it('runs surrogate context created internally', done => {
  const req = {};
  const spy = jest.spyOn(ContextRunnerImpl.prototype, 'run');
  oneOf([check('bar'), check('baz')])(req, {}, () => {
    expect(spy).toHaveBeenCalledTimes(3);
    // Calls 1 and 2 asserted by previous test
    expect(spy).toHaveBeenNthCalledWith(3, req, undefined);

    spy.mockRestore();
    done();
  });
});

it('runs in dry-run mode', async () => {
  const req = {};
  const spy = jest.spyOn(ContextRunnerImpl.prototype, 'run');
  await oneOf([check('bar'), check('baz')]).run(req, { dryRun: true });
  // Calls 1 and 2 asserted by previous test
  expect(spy).toHaveBeenNthCalledWith(3, req, { dryRun: true });
  spy.mockRestore();
});

it('passes unexpected errors down to other middlewares', done => {
  const error = new Error();
  const spy = jest.spyOn(ContextRunnerImpl.prototype, 'run').mockRejectedValue(error);

  oneOf([check('bar'), check('baz')])({}, {}, (err?: any) => {
    expect(err).toBe(error);

    spy.mockRestore();
    done();
  });
});

describe('with a list of chains', () => {
  it('sets a single error for the _error key', done => {
    const req: InternalRequest = {
      cookies: { foo: true, bar: 'def' },
    };

    oneOf([check('foo').isInt(), check('bar').isInt()])(req, {}, () => {
      const context = getOneOfContext(req);
      expect(context.errors).toMatchSnapshot();
      done();
    });
  });

  it('does not set an error if any chain succeeded', done => {
    const req: InternalRequest = {
      cookies: { foo: true, bar: 123 },
    };

    oneOf([check('foo').isInt(), check('bar').isInt()])(req, {}, () => {
      const context = getOneOfContext(req);
      expect(context.errors).toHaveLength(0);
      done();
    });
  });
});

describe('with a list of chain groups', () => {
  it('sets a single error for the _error key', done => {
    const req: InternalRequest = {
      cookies: { foo: true, bar: 'def', baz: 123 },
    };

    oneOf([[check('foo').isInt(), check('bar').isInt()], check('baz').isAlpha()])(req, {}, () => {
      const context = getOneOfContext(req);
      expect(context.errors).toMatchSnapshot();
      done();
    });
  });

  it('does not set an error if any chain group succeeded', done => {
    const req: InternalRequest = {
      cookies: { foo: true, bar: 'def', baz: 'qux' },
    };

    oneOf([[check('foo').isInt(), check('bar').isInt()], check('baz').isAlpha()])(req, {}, () => {
      const context = getOneOfContext(req);
      expect(context.errors).toHaveLength(0);
      done();
    });
  });

  it('stops running chains in a group when one of them has errors and request-level bail', done => {
    const req = {
      cookies: { foo: true, bar: 'def', baz: 'qux' },
    };

    const custom = jest.fn();
    oneOf([
      [check('foo').isInt().bail({ level: 'request' }), check('bar').custom(custom)],
      check('baz').isAlpha(),
    ])(req, {}, () => {
      expect(custom).not.toHaveBeenCalled();
      done();
    });
  });
});

describe('error message', () => {
  it('is "Invalid value(s)" by default', done => {
    const req: InternalRequest = {
      body: { foo: true },
    };

    oneOf([check('foo').isInt()])(req, {}, () => {
      const context = getOneOfContext(req);
      expect(context.errors[0]).toHaveProperty('msg', 'Invalid value(s)');
      done();
    });
  });

  it('can be a custom one', done => {
    const req: InternalRequest = {
      body: { foo: true },
    };

    oneOf([check('foo').isInt()], { message: 'not today' })(req, {}, () => {
      const context = getOneOfContext(req);
      expect(context.errors[0]).toHaveProperty('msg', 'not today');
      done();
    });
  });

  it('can be the return of a function', done => {
    const req: InternalRequest = {
      body: { foo: true },
    };

    const message = jest.fn(() => 'keep trying');
    oneOf([check('foo').isInt()], { message })(req, {}, () => {
      const context = getOneOfContext(req);
      expect(context.errors).toMatchSnapshot();
      done();
    });
  });
});

describe('should let the user to choose between multiple error types', () => {
  const errors: OneOfErrorType[] = ['grouped', 'flat'];
  it.each(errors)(`%s error type`, async errorType => {
    const req: InternalRequest = {
      body: { foo: true },
    };
    const options: OneOfOptions = {
      errorType,
    };

    await oneOf([check('foo').isString(), check('bar').isFloat()], options).run(req);
    const context = getOneOfContext(req);
    expect(context.errors).toMatchSnapshot();
  });

  it('leastErroredOnly error type', done => {
    const req: InternalRequest = {
      body: { foo: true, bar: 'bar' },
    };
    const options: OneOfOptions = {
      errorType: 'leastErroredOnly',
    };

    oneOf(
      [
        [check('foo').isFloat(), check('bar').isInt()],
        [check('foo').isString(), check('bar').isString()], // least errored
        [check('foo').isFloat(), check('bar').isBoolean()],
      ],
      options,
    )(req, {}, () => {
      const context = getOneOfContext(req);
      expect(context.errors).toMatchSnapshot();
      done();
    });
  });
});

describe('should default to grouped errorType', () => {
  it('when no options are provided', async () => {
    const req: InternalRequest = {
      body: { foo: true },
    };
    await oneOf([check('foo').isString(), check('bar').isFloat()]).run(req);
    const context = getOneOfContext(req);
    expect(context.errors[0]).toEqual(
      expect.objectContaining({
        type: 'alternative_grouped',
        nestedErrors: [expect.anything(), expect.anything()],
      }),
    );
  });

  it('when invalid error type is provided', async () => {
    const req: InternalRequest = {
      body: { foo: true },
    };
    await oneOf([check('foo').isString(), check('bar').isFloat()], {
      // @ts-ignore
      errorType: 'invalid error type',
    }).run(req);
    const context = getOneOfContext(req);
    expect(context.errors[0]).toEqual(
      expect.objectContaining({
        type: 'alternative_grouped',
        nestedErrors: [expect.anything(), expect.anything()],
      }),
    );
  });
});

describe('imperatively run oneOf', () => {
  it('sets errors in context when validation fails', async () => {
    const req: InternalRequest = {
      body: { foo: true },
    };

    const middleware = oneOf([check('foo').isInt()]);
    await middleware.run(req);

    const context = getOneOfContext(req);
    expect(context.errors.length).toEqual(1);
  });

  it('should throw an error if ContextRunner throws', async () => {
    const req: InternalRequest = {
      body: { foo: true },
    };
    const error = new Error();
    const spy = jest.spyOn(ContextRunnerImpl.prototype, 'run').mockRejectedValue(error);

    const middleware = oneOf([check('foo').isBoolean()]);
    await expect(middleware.run(req)).rejects.toThrow(error);
    spy.mockRestore();
  });

  it('sets no error in context when successful', async () => {
    const req: InternalRequest = {
      body: { foo: true },
    };

    const middleware = oneOf([check('foo').isBoolean()]);
    const result = await middleware.run(req);

    const context = getOneOfContext(req);
    expect(result).toBeInstanceOf(Result);
    expect(context.errors.length).toEqual(0);
  });
});
