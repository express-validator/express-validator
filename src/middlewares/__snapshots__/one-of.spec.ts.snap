// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error message can be the return of a function 1`] = `
Array [
  Object {
    "msg": "keep trying",
    "nestedErrors": Array [
      Array [
        Object {
          "location": "body",
          "msg": "Invalid value",
          "path": "foo",
          "type": "field",
          "value": true,
        },
      ],
    ],
    "type": "alternative_grouped",
  },
]
`;

exports[`should let the user to choose between multiple error types flat error type 1`] = `
Array [
  Object {
    "msg": "Invalid value(s)",
    "nestedErrors": Array [
      Object {
        "location": "body",
        "msg": "Invalid value",
        "path": "foo",
        "type": "field",
        "value": true,
      },
      Object {
        "location": "body",
        "msg": "Invalid value",
        "path": "bar",
        "type": "field",
        "value": undefined,
      },
    ],
    "type": "alternative",
  },
]
`;

exports[`should let the user to choose between multiple error types grouped error type 1`] = `
Array [
  Object {
    "msg": "Invalid value(s)",
    "nestedErrors": Array [
      Array [
        Object {
          "location": "body",
          "msg": "Invalid value",
          "path": "foo",
          "type": "field",
          "value": true,
        },
      ],
      Array [
        Object {
          "location": "body",
          "msg": "Invalid value",
          "path": "bar",
          "type": "field",
          "value": undefined,
        },
      ],
    ],
    "type": "alternative_grouped",
  },
]
`;

exports[`should let the user to choose between multiple error types least_errored error type 1`] = `
Array [
  Object {
    "msg": "Invalid value(s)",
    "nestedErrors": Array [
      Object {
        "location": "body",
        "msg": "Invalid value",
        "path": "foo",
        "type": "field",
        "value": true,
      },
    ],
    "type": "alternative",
  },
]
`;

exports[`with a list of chain groups sets a single error for the _error key 1`] = `
Array [
  Object {
    "msg": "Invalid value(s)",
    "nestedErrors": Array [
      Array [
        Object {
          "location": "cookies",
          "msg": "Invalid value",
          "path": "foo",
          "type": "field",
          "value": true,
        },
        Object {
          "location": "cookies",
          "msg": "Invalid value",
          "path": "bar",
          "type": "field",
          "value": "def",
        },
      ],
      Array [
        Object {
          "location": "cookies",
          "msg": "Invalid value",
          "path": "baz",
          "type": "field",
          "value": 123,
        },
      ],
    ],
    "type": "alternative_grouped",
  },
]
`;

exports[`with a list of chains sets a single error for the _error key 1`] = `
Array [
  Object {
    "msg": "Invalid value(s)",
    "nestedErrors": Array [
      Array [
        Object {
          "location": "cookies",
          "msg": "Invalid value",
          "path": "foo",
          "type": "field",
          "value": true,
        },
      ],
      Array [
        Object {
          "location": "cookies",
          "msg": "Invalid value",
          "path": "bar",
          "type": "field",
          "value": "def",
        },
      ],
    ],
    "type": "alternative_grouped",
  },
]
`;
